src/include/detail/ivf/index.h

```diff
@@ -109,16 +113,18 @@ int ivf_index(
     // be difficult to implement.  Even this algorithm is not trivial to
     // parallelize, because of the random access to the indices array.
     for (size_t i = 0; i < db.num_cols(); ++i) {
-      size_t bin = parts[i];
-      size_t ibin = indices[bin];
+      if (auto iter = deleted_ids_set.find(external_ids[i]); iter == deleted_ids_set.end()) {
```

Comment:
This is going to be expensive.  (log(size(deleted_ids) * size(db)).  But I don't immediately see an obvious linear algorithm (or parallelization).  But if size(deleted_ids) << size(db) it might not be horrible.

Reply:
Avoided adding complexity to ingestions that don't need `deleted_ids` by only checking the set when `deleted_ids` is not empty.

When we do need to check for `deleted_ids` we could try to have the `db` filtered once and then iterate over just the filtered vectors. I left this for a future PR. I am not sure if changing the loop order would give much performance benefit here but we can try to do a performance evaluation.